### ISSUE TITLE
Improve Schraffur

### ISSUE THREAD

**Issue by der-stefan** [der-stefan/OpenTopoMap#105]
Wohngebäude 45°, Nutzgebäude 90°

**Comment by max-dn** [comment]
Hatching in 45° for building = yes / residential / commercial (?) ... and 0° or 135° for all others? We could try... 
Maybe the other way: one hatching for building=barn, carport, bridge, stable... and one for each other.

The rule "Alle anderen Gebäude werden rechtwinklig zur längsten Gebäudeseite schraffiert." ("all other hatching 90° to the longest wall of the building") seems heavy, we don't have the rotation of a building. And it's ugly: All buildings will look like terraced houses.

**Comment by der-stefan** [comment]
Yes, I think this issue is very hard to implement. We would have to calculate the main direction of EVERY building. It would be only necessary for zoom 17...

**Comment by max-dn** [comment]
mapnik-de and the french style are the only mapnik maps with on-the-fly rotation, as far as I know. They use [some confusing queries](https://github.com/giggls/openstreetmap-carto-de/blob/master/project.mml#L624) to rotate the symbol for sports grounds... Maybe something like that, but easier, some of the variables there are only for estimating wether a pitch is a rectangle.

**Comment by der-stefan** [comment]
Wow, great! That is a good starting point...
By the way: Drawing sports ground as a rotated and scaled symbols (on zoom 13-15) would be nice, too! I received a lot of requests by users.

**Comment by max-dn** [comment]
The code for the soccer fields needs a rectangular field. Sometimes we have more than 4 corners mapped and thats why the code uses a [simplification](https://postgis.net/docs/ST_SimplifyPreserveTopology.html) of the way to calculate the rotation of the yard lines. We need something similar to get the longest wall without smaller details. 

I rendered such a simplification for all buildings: 
![simplified buildings](http://geo.dianacht.de/bilder/otm_simplified_buildings.png)

That works fine for rectangular buildings, but not for complex architecture like the one between Heß- and Gabelsbergerstraße or L-shaped buildings (corner of Gabelsberger/Arcis). So maybe this code will work under the assumption, that not-residential buildings are rectangular (ok, most barns are rectangular... offices and museums and hospitals are not...)

(It looks worse than it really is. For example the L-building Theresien/Luisenstr: It is rather deformed but the direction of the longest way in the left image is correct)

**Comment by der-stefan** [comment]
The simplified buildings are not for being rendered, but only for the calculation of the hatching, right?

**Comment by max-dn** [comment]
Yes, I just needed a feeling, what this "ST_SimplifyPreserveTopology" is doing with an area. The hatching could be 90° to the longest way of a simplified shape.

**Comment by max-dn** [comment]
Just a question: Is there any easy way to get rotated hatches with mapnik?
If I take a small svg like mapnik/symbols-otm/building_z17.svg i get hatches with 45°. If I rotate this svg by 20°, each 5x5px tile is rotated and then these tiles are stitched together:

![rotated building_z17.svg](https://geo.dianacht.de/bilder/otm_hatchrotation.png)

If there is no easy way, I could make a svg with 400x400px and rotate this (and hope no building is larger than 400px...)

**Comment by der-stefan** [comment]
Yes, I think it is the best way to create a very large svg pattern. It should be no problem to have a pattern of (1000x1000) px² ;-)

**Comment by der-stefan** [comment]
Revoke: Mapnik is stupid. It always stitches the pattern together, so you **can not** avoid ugly joints. :-1:
